File: runtime/cpp/titan-ast-runtime-lib/FollowFilterBacktrackingBottomUpAstAutomata.cpp
//
// Created by tian wei jun on 2024/1/16.
//

#include "FollowFilterBacktrackingBottomUpAstAutomata.h"
FollowFilterBacktrackingBottomUpAstAutomata::
    FollowFilterBacktrackingBottomUpAstAutomata(
        const SyntaxDfa *astDfa, const Grammar *startGrammar,
        Grammar **innerGrammars, int countOfInnerGrammars,
        const std::unordered_map<
            const Grammar *,
            std::unordered_set<const Grammar *, PtrGrammarContentHash,
                               PtrGrammarContentEq> *,
            PtrGrammarContentHash, PtrGrammarContentEq> *nonterminalFollowMap,
        const Grammar *eofGrammar)
    : BacktrackingBottomUpAstAutomata(astDfa, startGrammar, innerGrammars,
                                      countOfInnerGrammars),
      nonterminalFollowMap(nonterminalFollowMap), eofGrammar(eofGrammar) {}

void FollowFilterBacktrackingBottomUpAstAutomata::reduceBottomUpBranch(
    BacktrackingBottomUpBranch *bottomUpBranch) {
  ReducingSymbol *topReducingSymbol = bottomUpBranch->reducingSymbols.back();
  const SyntaxDfaState *currentDfaState = topReducingSymbol->currentDfaState;
  if (!currentDfaState->closingProductionRules.empty()) {

    tokenReducingSymbolInputStream.nextReadIndex =
        topReducingSymbol->endIndexOfToken + 1;

    const Grammar *terminalOfNextToken = nullptr;     // 下一个token的语法名字
    if (tokenReducingSymbolInputStream.hasReadAll()) {// token读完了,相当于eof
      terminalOfNextToken = eofGrammar;
    } else {
      terminalOfNextToken = tokenReducingSymbolInputStream.read()->terminal;
    }

    auto &closingProductionRules = currentDfaState->closingProductionRules;
    for (auto closingProductionRule : closingProductionRules) {

      const Grammar *nonterminal = closingProductionRule->grammar;
      auto follow = nonterminalFollowMap->find(nonterminal)->second;

      if (follow->find(terminalOfNextToken) != follow->end()) {
        doReduce(bottomUpBranch, closingProductionRule);
      }
    }
  }
}
AstAutomataType FollowFilterBacktrackingBottomUpAstAutomata::getType() {
  return AstAutomataType::FOLLOW_FILTER_BACKTRACKING_BOTTOM_UP_AST_AUTOMATA;
}

File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/FollowFilterBacktrackingBottomUpAstAutomata.java
package titan.ast.runtime;

import java.util.Map;
import java.util.Set;

/**
 * 按照可回溯方式构造ast的自动机.
 *
 * @author tian wei jun
 */
class FollowFilterBacktrackingBottomUpAstAutomata extends BacktrackingBottomUpAstAutomata {
  Map<Grammar, Set<Grammar>> nonterminalFollowMap;
  Grammar eof;

  FollowFilterBacktrackingBottomUpAstAutomata(
      SyntaxDfa astDfa,
      Grammar startGrammar,
      Map<Grammar, Set<Grammar>> nonterminalFollowMap,
      Grammar eof) {
    super(astDfa, startGrammar);
    this.nonterminalFollowMap = nonterminalFollowMap;
    this.eof = eof;
  }

  @Override
  public AstAutomataType getType() {
    return AstAutomataType.FOLLOW_FILTER_BACKTRACKING_BOTTOM_UP_AST_AUTOMATA;
  }

  @Override
  void reduceBottomUpBranch(BacktrackingBottomUpBranch bottomUpBranch) {
    ReducingSymbol topReducingSymbol = bottomUpBranch.reducingSymbols.getLast();
    tokenReducingSymbolInputStream.nextReadIndex = topReducingSymbol.endIndexOfToken + 1;

    Grammar terminalOfNextToken = null; // 下一个token的语法名字
    if (tokenReducingSymbolInputStream.hasReadAll()) { // token读完了,相当于eof
      terminalOfNextToken = eof;
    } else {
      terminalOfNextToken = tokenReducingSymbolInputStream.read().terminal;
    }
    // 还有token没有规约，判断token是不是在follow集里决定是否规约
    for (ProductionRule closingProductionRule :
        topReducingSymbol.currentDfaState.closingProductionRules) {
      Grammar nonterminal = closingProductionRule.grammar;
      Set<Grammar> follow = nonterminalFollowMap.get(nonterminal);
      if (follow.contains(terminalOfNextToken)) {
        doReduce(bottomUpBranch, closingProductionRule);
      }
    }
  }
}
